runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -343,11 +343,11 @@ public void clear(K key, W window) {
         }
 
         private void prepareStateBackend(K key) {
-          // Key for state request is shipped already encoded as ByteString,
-          // this is mostly a wrapping with ByteBuffer. We still follow the
-          // usual key encoding procedure.
-          // final ByteBuffer encodedKey = FlinkKeyUtils.encodeKey(key, keyCoder);
-          final ByteBuffer encodedKey = ByteBuffer.wrap(key.toByteArray());
+          // Key for state request is shipped already encoded as ByteString, but it is
```

Comment:
In which situation is it NESTED? We made the change originally because it wasn't prefixed, so trying to understand what changed.

Reply:
It is always `NESTED` although it is not always apparent because not all coders use a length prefix. To correctly match the Flink Runner's key encoding, we need to remove the length prefix, i.e. convert `NESTED` to `OUTER` context.

Reply:
Based on https://github.com/apache/beam/pull/8472/files there shouldn't be a length prefix?

Reply:
Yes, but that is incorrect. It worked before because we were applying a double nested encoding.

Reply:
I should say, the bug only manifests together with timers which access state, because there the key encoding will not match between the two.

Reply:
The timer test should be adjusted accordingly to catch that.

Reply:
There is something that I'm missing. State and timers work as of release 2.14 with Flink 1.8. What has caused it to break?

Reply:
https://gist.github.com/tweise/e1d95369234a52b12f90794f6032f039

That's working on 2.14 and also on master as of 4460f03cefbde5cf66c797408439321de2f40692

Reply:
Here is why this works for Flink <= 1.8 but definitely needs to be fixed moving forward:

Keys for state requests are encoded incorrectly for the Flink Runner's state backend because they may contain a length prefix, which is not in lines with how the key is encoded for data partitioning. This causes state to be written to the wrong key group. However, the encoding scheme is consistent, so reading from state in `ProcessElement` or `OnTimer` methods works correctly. Also when inter-playing with timers. However, there are potential issues with checkpoints/savepoints due to this.

In Flink 1.9, the key encoding mismatch is visible because reading state from a key not part of the partition _can_ return a `null` value. That is, not always, likely due to the state compaction logic. This explains why `PortableStateExecutionTest` works correctly but `PortableTimersExecutionTest` does not. That said, I think we should merge this fix ASAP for the upcoming release.

Reply:
OK, thanks for the RCA. I will take this PR for a test drive and report back. (Just FYI no checkpointing issues with the current encoding on Flink 1.8).